src/gtk/toga_gtk/widgets/canvas.py

```diff
@@ -33,142 +27,158 @@ def create(self):
 
         self.native = Gtk.DrawingArea()
         self.native.interface = self.interface
-        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, self.native.get_allocated_width(),
-                                     self.native.get_allocated_height())
-        self.native.context = cairo.Context(surface)
-        self.native.font = None
 
-    def set_on_draw(self, handler):
-        self.native.connect('draw', handler)
+    def create_draw_callback(self, root_context):
+        """Creates a draw callback
 
-    def set_context(self, context):
-        self.native.context = context
+        Gtk+ uses a drawing callback to draw on a DrawingArea. Assignment of the
+        callback function creates a Gtk+ canvas and Gtk+ context automatically
+        using the canvas and native_context function arguments. This method
+        traverses through the full tree of drawing objects and uses the
+        callback to call each one.
 
-    def line_width(self, width=2.0):
-        self.native.context.set_line_width(width)
+        """
 
-    def fill_style(self, color=None):
-        if color is not None:
-            num = re.search('^rgba\((\d*\.?\d*), (\d*\.?\d*), (\d*\.?\d*), (\d*\.?\d*)\)$', color)
-            if num is not None:
-                #  Convert RGB values to be a float between 0 and 1
-                r = float(num.group(1)) / 255
-                g = float(num.group(2)) / 255
-                b = float(num.group(3)) / 255
-                a = float(num.group(4))
-                self.native.context.set_source_rgba(r, g, b, a)
-            else:
-                pass
-                # Support future colosseum versions
-                # for named_color, rgb in colors.NAMED_COLOR.items():
-                #     if named_color == color:
-                #         exec('self.native.set_source_' + str(rgb))
-        else:
-            # set color to black
-            self.native.context.set_source_rgba(0, 0, 0, 1)
+        def draw_callback(canvas, context):
+            for drawing_object in root_context.drawing_objects:
+                drawing_object(self, native_context=context)
+
+        self.native.connect("draw", draw_callback)
+
+    def redraw(self, root_context):
+        pass
 
-    def stroke_style(self, color=None):
-        self.fill_style(color)
+    # Basic paths
 
-    def new_path(self):
-        self.native.context.new_path()
+    def new_path(self, native_context):
+        native_context.new_path()
 
-    def close_path(self):
-        self.native.context.close_path()
+    def closed_path(self, x, y, native_context):
```

Comment:
Ok, changed this. I know you aren't a staticmethod fan, but I get lots of warnings without it - so they all got staticmethod decorators as well.

Reply:
Nevermind. After looking at this more, I don't think this will work. The interface calls `impl.move_to()`, for example, so the impl gets passed to the self argument, which is what we want.